llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18713,6 +18729,12 @@ SDValue DAGCombiner::visitSINT_TO_FP(SDNode *N) {
   if (SDValue FTrunc = foldFPToIntToFP(N, DL, DAG, TLI))
     return FTrunc;
 
+  // fold (sint_to_fp (trunc nsw x)) -> (sint_to_fp x)
+  if (N0.getOpcode() == ISD::TRUNCATE && N0->getFlags().hasNoSignedWrap() &&
+      TLI.isTypeDesirableForOp(ISD::SINT_TO_FP,
```

Comment:
Makes me wonder which conversion variants we actually have h/w support for on the GPU.
PTX spec effectively claims that (nearly) anything to anything is supported, but I don't know whether we should LLVM do some of the job, or if we can always let ptxas do it.
It's not an issue for this patch, but a general though that we may want to provide a custom `TLI.isTypeDesirableForOp` for NVPTX. AFAICT we're currently running with the default which accepts all legal types.